ddspipe_core/include/ddspipe_core/efficiency/payload/PayloadPoolMediator.hpp

```diff
@@ -0,0 +1,141 @@
+#include <fastdds/rtps/history/IPayloadPool.h>
+
+#include <ddspipe_core/efficiency/payload/PayloadPool.hpp>
+#include <ddspipe_core/types/data/RtpsPayloadData.hpp>
+
+namespace eprosima {
+namespace ddspipe {
+namespace core {
+
+/**
+ * This class works as a mediator between the FastPayloadPool, the TopicDataType, and the DDS DataWriter write
+ * function. Its purpose is to save a copy between the DataWriter and the FastPayloadPool.
+ */
+class PayloadPoolMediator : public fastrtps::rtps::IPayloadPool
+{
+public:
+
+    /**
+     * TODO
+     */
+    PayloadPoolMediator(
+            const std::shared_ptr<PayloadPool>& payload_pool);
+
+    /**
+     * TODO
+     */
+    bool write(
+            fastdds::dds::DataWriter* writer,
+            types::RtpsPayloadData* data);
+
+    /**
+     * TODO
+     */
+    bool write(
+            fastdds::dds::DataWriter* writer,
+            types::RtpsPayloadData* data,
+            fastrtps::rtps::WriteParams& params);
+
+    /**
+     * TODO
+     */
+    utils::ReturnCode write(
+            fastdds::dds::DataWriter* writer,
+            types::RtpsPayloadData* data,
+            const fastrtps::rtps::InstanceHandle_t& handle);
+
+    /**
+     * TODO
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool get_payload(
+            uint32_t size,
+            fastrtps::rtps::CacheChange_t& cache_change) override;
+
+    /**
+     * TODO
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool get_payload(
+            fastrtps::rtps::SerializedPayload_t& data,
+            fastrtps::rtps::IPayloadPool*& data_owner,
+            fastrtps::rtps::CacheChange_t& cache_change) override;
+
+    /**
+     * TODO
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool release_payload(
+            fastrtps::rtps::CacheChange_t& cache_change) override;
+
+    /**
+     * Instead of reserving a new payload, since we know the payload and its owner,
+     * we redirect it to the other FastPayloadPool function.
+     *
+     * @param size size of the new chunk of data
+     * @param payload object to store the new data
+     *
+     * @return true if everything OK
+     * @return false if something went wrong
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool get_payload(
+            uint32_t size,
+            types::Payload& payload);
+
+    /**
+     * TODO
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool get_payload(
+            const types::Payload& src_payload,
+            fastrtps::rtps::IPayloadPool*& data_owner,
+            types::Payload& dst_payload);
+
+    /**
+     * TODO
+     */
+    DDSPIPE_CORE_DllAPI
+    virtual bool release_payload(
+            types::Payload& payload);
+
+protected:
+
+    //! Mutex to prevent simultaneous calls to write
+    std::mutex mutex_;
+
+    //! A pointer to the data_ we are trying to write with the write operation.
+    types::RtpsPayloadData* data_;
+
+    //! TODO
+    const std::shared_ptr<PayloadPool>& payload_pool_;
```

Comment:
Why not IPayloadPool?

Reply:
The PayloadPoolMediator calls FastPayloadPool methods that are declared as purely virtual in PayloadPool.